pkg/controller/v1beta1/inferenceservice/utils/utils.go

```diff
@@ -112,14 +103,57 @@ func MergeRuntimeContainers(runtimeContainer *v1.Container, predictorContainer *
 		coreContainer.Name = runtimeContainerName
 	}
 
-	argCopy = append(argCopy, predictorContainer.Args...)
-	envCopy = append(envCopy, predictorContainer.Env...)
-
-	coreContainer.Args = argCopy
-	coreContainer.Env = envCopy
+	// Args and Env will be combined instead of overridden
+	argCopy := make([]string, len(runtimeContainer.Args))
+	copy(argCopy, runtimeContainer.Args)
+	coreContainer.Args = append(argCopy, predictorContainer.Args...)
+	coreContainer.Env = MergeEnvs(runtimeContainer.Env, predictorContainer.Env)
 
 	return &coreContainer, nil
+}
 
+// MergeRuntimeVolumes Merge the predictor Volumes list with the runtime Volumes list, allowing users
+// to override/augment runtime volume settings from the predictor spec.
+func MergeRuntimeVolumes(runtimeVolumes []v1.Volume, predictorVolumes []v1.Volume) []v1.Volume {
+	if len(runtimeVolumes) == 0 {
+		return predictorVolumes
+	} else if len(predictorVolumes) == 0 {
+		return runtimeVolumes
+	}
+	vols := append(([]v1.Volume)(nil), runtimeVolumes...)
+outer:
+	for ip, _ := range predictorVolumes {
+		for ir, _ := range vols {
+			if predictorVolumes[ip].Name == vols[ir].Name {
+				vols[ir] = predictorVolumes[ip]
+				continue outer
+			}
+		}
+		vols = append(vols, predictorVolumes[ip])
+	}
+	return vols
+}
+
+// MergeEnvs Merge the predictor Env list with the runtime Env list, allowing users
+// to override/augment environment variables from the predictor spec.
+func MergeEnvs(runtimeEnv []v1.EnvVar, predictorEnv []v1.EnvVar) []v1.EnvVar {
+	if len(runtimeEnv) == 0 {
+		return predictorEnv
+	} else if len(predictorEnv) == 0 {
+		return runtimeEnv
+	}
+	env := append(([]v1.EnvVar)(nil), runtimeEnv...)
```

Comment:
If you want to ensure a single allocation, can extend the slice with capacity in case each predictor entry is added:
```suggestion
	env := make([]v1.Volume, len(runtimeEnv), len(runtimeEnv)+len(predictorEnv))
	copy(env, runtimeEnv)
```

Reply:
Same answer :)